File: leetcode/implement-trie-prefix-tree.java
// https://leetcode.com/problems/implement-trie-prefix-tree

class Trie {
    private boolean wordEnd = false;
	private HashMap<Character, Trie> child = new HashMap<Character, Trie>();

    /** Initialize your data structure here. */
    public Trie() {}
    
    /** Inserts a word into the trie. */
    public void insert(String s) {
		Trie node = this;
		for (int i = 0, n = s.length(); i < n; ++i) {
            // if (!node.child.containsKey(s.charAt(i)))
            //     node.child.put(s.charAt(i), new Trie());
			node.child.putIfAbsent(s.charAt(i), new Trie());
			node = node.child.get(s.charAt(i));
		}
		node.wordEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String s) {
        // return true;
		Trie node = this;
		for (int i = 0, n = s.length(); i < n; ++i) {
			if (!node.child.containsKey(s.charAt(i)))
				return false;
			node = node.child.get(s.charAt(i));
		}
		return node.wordEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String s) {
        // return true;
		Trie node = this;
		for (int i = 0, n = s.length(); i < n; ++i) {
			if (!node.child.containsKey(s.charAt(i)))
				return false;
			node = node.child.get(s.charAt(i));
		}
		return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: leetcode/implement-trie-prefix-tree.cpp
class Trie {
    Trie* child[26];
    bool isTerminal;
public:
    Trie() {
        for (int i = 0; i < 26; ++i)
            child[i] = nullptr;
        isTerminal = false;
    }
    
    void insert(string word, int idx = 0) {
        if (idx >= word.size()) return;
        char c = word[idx];
        if (!child[c-'a']) child[c-'a'] = new Trie();
        Trie* node = child[c-'a'];
        node->isTerminal |= idx == word.size()-1;
        node->insert(word, idx+1);
    }
    
    bool search(string word) {
        Trie* node = child[word[0]-'a'];
        for (int i = 1; i < word.size(); ++i) {
            if (!node) return false;
            node = node->child[word[i]-'a'];
        }
        if (!node) return false;
        return node->isTerminal;
    }
    
    bool startsWith(string word) {
        Trie* node = child[word[0]-'a'];
        for (int i = 1; i < word.size(); ++i) {
            if (!node) return false;
            node = node->child[word[i]-'a'];
        }
        if (!node) return false;
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */